### PR TITLE
RFC: absolutely minimal way to close #18823

### LINKED ISSUE #18823: Julep: Revised iteration protocol
We discussed at JuliaCon what to do about #8149. I'm opening this issue to capture the specific design proposed at JuliaCon (and so that others can make corrections about things I misremembered).
# The new iteration protocol

In the new iteration protocol, there's only one function to be overloaded:

```
step(it, state=...)::Nullable
```

with for loops getting lowered as:

```
for i in it
    #= loop body =#
end
```

equivalent to

```
x = step(it)
while !isnull(x)
    i, state = unsafe_get(x)
    #= loop body =#
    x = step(it, state)
end
```
# Implementation

This takes care of iterators that don't know whether they're done until trying (just return `Nullable{typeof(state}()`). One concern with this approach (and approaches like it) was whether there would be an extra allocation and branch, which would be an unacceptable performance regression. (Added in EDIT by @timholy: https://github.com/JuliaLang/julia/issues/16035#issue-150794573 and #9080.) The conclusion we came to was that the compiler should be able to optimize this out. To make sure that this happens, we may want to use a slightly different lowering, such as:

```
1:
x = step(it)
isnull(x) ? goto 4 : goto 3
2:
x = step(it, state)
isnull(x) ? goto 4 : goto 3
3:
#= loop body =#
goto 2:
4:
#= rest of function =#
```

The motivation for structuring things that way, is that simple step functions where performance really matters probably look something like:

```
function step(it, state)
if state > length(it)
Nullable{Tuple{T,Int}}()
else
Nullable{Tuple{T,Int}}((it[state], state+1))
end
end
```

which after inlining, should make this a fairly straightforward jump threading optimization (if the optimization doesn't happen we should figure out why and fix that).
# Migration

Another nice property of this proposal is that it's backwards compatible with the existing scheme. The fallback method is simply:

```
@inline function step(it,state=start(it))
done(it) ? Nullable{Tuple{Any,typeof(state)}}() : Nullable{Tuple{Any, typeof(state)}}(next(it, state))
end
```

Now, the `Nullable{Any,}` in there is not ideal, but forced inlining together with allocation elision should be able to handle that just fine (otherwise we could also do `Nullable{eltype(it),...}`, which should work also, but has slightly greater risk of breaking).

cc @timholy @JeffBezanson - Please add anything I missed

### PR DESCRIPTION
**Pull request by StefanKarpinski**
I think that this would be far less disruptive than the proposal in https://github.com/JuliaLang/julia/issues/18823, not to mention that it's getting too late to make such a huge change to the iteration protocol at this point. This allows iterators that don't know they're done until they've tried to return `nothing` from the `next` method instead of a `value, state` tuple and return early that way. There's no impact on iterators that don't need to do that and there's no impact on code that uses the iteration protocol to process generic iterators as long as they don't need to work with "unforseeable iterators". So any code that works for arrays, strings, etc. needs no changing. Generic code that needs to work with these kinds of iterators can be fixed gradually, as we encounter it failing in the wild. What does need to be done in addition to this minimal changes is to make all the iterators in Base that should use this functionality actually use it.

This approach does have a few issues, but which I believe it shares with the proposal in #18823:

* The generic `isempty` definition seems wrong now since it checks `done(itr, state)` which for this kind of iterator will return `false`, yet the following call to `next` can return `nothing` leaving one with no values, despite `isempty` being false. Perhaps this is ok if we document that for such iterators the `isempty` predicate can be unreliable: the iterator may not _know_ that it's empty, yet not return any valules.

* There are probably other generic uses of `done` like the definition of `isempty` that would have similar problems.

* @JeffBezanson has expressed some concerns about the expansion of code involved in the new lowering. It doesn't emit that much extra code, but for loops are very common so it could still be a problem.

### CODE REVIEW

**Comment by StefanKarpinski** [comment]
* Win32 run succeeded but then timed out on some network stuff
* Win64 timed out on early network stuff
* Travis Linux runs both succeeded

File: src/julia-syntax.scm [hunk 1]
```diff
@@ -1646,8 +1647,9 @@
                  ,@(if (and (not outer?) (or *depwarn* *deperror*))
                        (map (lambda (v) `(warn-if-existing ,v)) (lhs-vars lhs))
                        '())
-                 ,(lower-tuple-assignment (list lhs state)
-                                          `(call (top next) ,coll ,state))
+                 (= ,nxt (call (top next) ,coll ,state))
+                 (if (call (core ===) ,nxt (core nothing)) (break))
```
Comment (stevengj) [inline src/julia-syntax.scm]: I assume this check gets optimized out in the common case of a `next` function that always returns a tuple?

Hopefully it gets optimized out even for iteration over a type-unstable container, e.g. where `next` returns `Tuple{ ...inference fails... , MyState}`?

**Comment by JeffBezanson** [comment]
From triage: we'll try to implement the full new iteration protocol and see how far we get. The protocol can include both `iterate` and optionally `done`. If an iterator is able to know that it's done without advancing, it can implement `done`, and otherwise not implement it.

**Comment by StefanKarpinski** [comment]
Since I goaded Keno and Jameson into opening https://github.com/JuliaLang/julia/pull/25261, my work is done here.

**Pull request closed without merging**